### PR TITLE
Fix float16 and mixed precision for crf

### LINKED ISSUE #2030: float16 support for text/crf
text/crf.py doesn't support float16 training or TF Lite converter.
According to the error information, the type of  masks In line 256 and line 294 is tf.float32:
```
masks = tf.sequence_mask(
        sequence_lengths, maxlen=tf.shape(tag_indices)[1], dtype=tf.float32
    )
```
which raises error when doling multiply op, like in line 259: 
```unary_scores = tf.reduce_sum(unary_scores * masks, 1)```

**Which API type would this fall under (layer, metric, optimizer, etc.)**
text
image

### PR DESCRIPTION
**Pull request by WindQAQ**
# Description

Brief Description of the PR:
Fix float16/mixed precision of CRF

Fixes #2030

## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*  Use mixed precision mixture and parameterized on float16

### CODE REVIEW

File: tensorflow_addons/text/crf.py [hunk 1]
```diff
@@ -209,7 +213,7 @@ def crf_log_likelihood(
         transition_params = tf.Variable(
             initializer([num_tags, num_tags]), "transitions"
         )
-
+    transition_params = tf.cast(transition_params, inputs.dtype)
```
Comment (bhack) [inline tensorflow_addons/text/crf.py:216]: I think that by a computational point of view probably it is quite the same so do you think that it makes sense to put in the `else` branch?

Comment (WindQAQ) [reply tensorflow_addons/text/crf.py:216]: I think current way is much safer when it talks about mixed precision. I figure out that this casting will maks the returned `transition_params` have different type from the input one.

BTW, from the aspect of design pattern, I do not think this function should accept `None` of `transition_params` as input, creates it and returns it. Users should prepare the `transition_params` in advance and call the function. This is how our `layers/crf_test.py` does.

Comment (bhack) [reply tensorflow_addons/text/crf.py:216]: Yes probably.
I only meant that Variabile has dtype if we will maintain the if.

Comment (bhack) [reply tensorflow_addons/text/crf.py:216]: @zara44 Just an opinion from @facaiy or @Squadrick about removing `None` case.

Comment (zara44) [reply tensorflow_addons/text/crf.py:216]: @Squadrick any chance you can have a look at this? :)

Comment (zara44) [reply tensorflow_addons/text/crf.py:216]: @facaiy are you able to look into this merge request? Thank you!

Comment (zara44) [reply tensorflow_addons/text/crf.py:216]: @seanpmorgan are you able to review this?

Comment (WindQAQ) [reply tensorflow_addons/text/crf.py:216]: @bhack can you approve this first? I believe they are now busy at their work, so may not response recently. Seems that this PR might be helpful for the community, I would think we can merge this first as a quick fix for mixed precision. Any suggestion change would also be really appreciated. Thanks!

Comment (bhack) [reply tensorflow_addons/text/crf.py:216]: Ok can you add an inline Todo to re-evaluate It later?

Comment (WindQAQ) [reply tensorflow_addons/text/crf.py:217]: Done 😃
